frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2672,50 +2672,73 @@ const
 std::vector<ID> findUsedImportedIds(Context* context, const Scope* scope) {
   auto result = resolveVisibilityStmts(context, scope);
   std::vector<ID> ids;
+  std::set<ID> idsSet;
 
   if (result == nullptr) return ids;
 
-  for (const auto& r : result->visibilityClauses()) {
-    ids.push_back(r.scope()->id());
+  // TODO: should this run the scope resolver?
+  // doing so would be necessary to find a submodule that is "used"
+  // through qualified access e.g.
+  // module Main {
+  //   module SubModule { proc foo() { } }
+  //   SubModule.foo();
+  // }
+
+  for (const VisibilitySymbols& vs : result->visibilityClauses()) {
+    if (vs.kind() == VisibilitySymbols::SYMBOL_ONLY ||
+        vs.kind() == VisibilitySymbols::ALL_CONTENTS ||
+        vs.kind() == VisibilitySymbols::CONTENTS_EXCEPT) {
+      // in these cases, the module/enum being used is scope()->id()
+      ID id = vs.scope()->id();
+      auto p = idsSet.insert(id);
+      if (p.second) { ids.push_back(id); }
+    } else if (vs.kind() == VisibilitySymbols::ONLY_CONTENTS) {
+      // could be
+      //  import Module.Submodule
+      // or
+      //  import Module.function
+      // or
+      //  use Module only function
+      // Here, we are only interested in the first case.
+      for (const auto& pair : vs.names()) {
+        UniqueString declName = pair.first;
+        IdAndFlags::Flags filter = 0; // match everything
+        IdAndFlags::FlagSet excludeFilter;
+        std::vector<BorrowedIdsWithName> matches;
+        vs.scope()->lookupInScope(declName, matches, filter, excludeFilter);
+        for (const auto& bid : matches) {
+          for (const auto& id : bid) {
+            if (parsing::idIsModule(context, id)) {
+              auto p = idsSet.insert(id);
+              if (p.second) { ids.push_back(id); }
+            }
+          }
+        }
```

Comment:
I think having to do this is unfortunate. The core issue is that we "forgot" what the original visibility statement looked like, which makes it hard to determine what 'thing' is being imported. After working with the LSP and processing visibility statements, I am a big fan of the 'normalized' form of visibility clauses. However, in this case, it seems to omit too much.

I'm skeptical of adding a new 'normalized' visibility statement mode that communicates that a child module is being imported. Brainstorming: could we add some kind of flag?

Another question I have is about the semantics of module dependencies. Isn't there an inconsistency? Consider:

* ```Chapel
   use Parent;
   ```
   The above won't consider 'Child' to be a 'used or imported' according to this logic here. However, if we (seemingly) restrict the statement:
* ```Chapel
  use Parent only Child;
  ```
   This includes an `ONLY_CONTENTS`, which will find `Child`. But in terms of normalized visibility statements, this is indistinguishable from...
* ```Chapel
  import Parent.{Child};
  ```
  
I confess I don't know the semantics of Chapel's resolution order, but my intuition here is that a visibility statement does _not_ bring in a module dependency **unless the module is used**. Thus, `use Parent` doesn't explicitly depend on `Child` (as now), and nor does `use Parent only Child` nor `import Parent.{Child}`. On the one hand, this violates the name of this function, which specifically asks for used and imported modules -- and `import Parent.Child` seemingly imports `Child`, as does `import Parent.{Child}`. But on the other hand, previous things you've said seemed to suggest that neither `chpl` nor `chpldoc` care about used/imported modules, but their mentions -- for dependencies? A recent PR also removed this need from `chapel-py`. Perhaps we can do away with this incarnation of the function, as well as this special logic?

This would require doing scope resolution like your above TODO suggests.

Reply:
> One alternative approach would be to have ResolvedVisibilityScope contain a field that is just a vector of module IDs for modules that are used or imported. That's just arguably moving the computation somewhere else. It would be worth it, though, if we think it's simpler to compute along with the rest of ResolvedVisibilityScope.

I've done this.